includes/class-wc-gateway-ppec-checkout-handler.php

```diff
@@ -476,6 +479,48 @@ public function maybe_return_from_paypal() {
 		}
 	}
 
+	/**
+	 * Updates shipping and billing addresses.
+	 *
+	 * Retrieves shipping and billing addresses from PayPal session.
+	 * This should be done prior to generating order confirmation so
+	 * local taxes can be calculated and displayed to customer.
+	 *
+	 * @since 1.6.2.
+	 *
+	 * @param string $token Token
+	 *
+	 * @return void
+	 */
+	private function update_customer_addresses_from_paypal( $token ) {
+		// Get the buyer details from PayPal.
+		try {
+			$checkout_details = $this->get_checkout_details( $token );
+		} catch ( PayPal_API_Exception $e ) {
+			wc_add_notice( $e->getMessage(), 'error' );
+			return;
+		}
+		$shipping_details = $this->get_mapped_shipping_address( $checkout_details );
+		$billing_details  = $this->get_mapped_billing_address( $checkout_details );
+
+		$customer = WC()->customer;
+
+		// Update billing/shipping addresses.
```

Comment:
Not needed for this PR but just for a future thought, I wonder if a method that wraps this functionality could DRY up this code quite a bit

Reply:
@c-shultz @laurendavissmith have grep-ed the code and found no other place that was using this so I thought that it is OK for one time. When I think about it now I think that I would have explored using `set_props`  from `WC_Data` https://docs.woocommerce.com/wp-content/images/wc-apidocs/source-class-WC_Data.html#589-622 
`WC_Customer` is a subclass of `WC_Data`. `set_props` uses a nice reflection trick that makes the `set_props` and this code equivalent.

Reply:
@budzanowski clever, i like that even better

Reply:
That's a good idea!

@budzanowski, something along the lines of how it's used here? - https://github.com/woocommerce/woocommerce/blob/3fa3830b6ceeaafda7eab8f5fa3534026913bd71/includes/class-wc-ajax.php#L293

That way, we can avoid as much repetition and the multiple "set" calls.

Reply:
I was thinking about something along the lines of:

```php
	private function update_customer_addresses_from_paypal( $token ) {
		// Get the buyer details from PayPal.
		try {
			$checkout_details = $this->get_checkout_details( $token );
		} catch ( PayPal_API_Exception $e ) {
			wc_add_notice( $e->getMessage(), 'error' );
			return;
		}
		$shipping = $this->get_mapped_shipping_address( $checkout_details );
		$billing  = $this->get_mapped_billing_address( $checkout_details );
		$props = [];

		foreach(['address_1', 'address_2', 'city', 'postcode', 'state', 'country'] as $key ) {
			$props['shipping_' . $key] = $shipping[ $key ];
			$props['billing_' . $key]  = $billing[ $key ];
		}

		WC()->customer->set_props( $props );
	}
```

Reply:
I don't have a very strong opinion that this is the best way. I just think that it is nicely short :)